skada/_reweight.py

```diff
@@ -669,3 +674,223 @@ def KLIEP(
+
+        def jac(alpha):
+            return P @ alpha.reshape(-1, 1) + q
+
+        B_beta = 10
+        eps = B_beta / (4 * np.sqrt(m))
+
+        constraints = LinearConstraint(
+            np.vstack([
+                R,
+                np.sum(R, axis=0, keepdims=True)
+            ]),
+            lb=np.concatenate([
+                np.zeros((R.shape[0])),
+                [m * (1 - eps)]
+            ]),
+            ub=np.concatenate([
+                B_beta * np.ones((R.shape[0])),
+                [m * (1 + eps)]
+            ])
+        )
+
+        results = minimize(
+            func,
+            x0=np.ones(R.shape[1]),
+            method="SLSQP",
+            jac=jac,
+            constraints=constraints,
+            tol=self.tol,
+            options={"maxiter": self.max_iter}
+        )
+
+        alpha = results.x
+        beta = R @ alpha
+        weights = beta.flatten()
+
+        return weights
+
+    def fit(self, X, y, sample_domain=None, **kwargs):
+        """Fit adaptation parameters.
+
+        Parameters
+        ----------
+        X : array-like, shape (n_samples, n_features)
+            The source data.
+        y : array-like, shape (n_samples,)
+            The source labels.
+        sample_domain : array-like, shape (n_samples,)
+            The domain labels (same as sample_domain).
+
+        Returns
+        -------
+        self : object
+            Returns self.
+        """
+        X, y, sample_domain = check_X_y_domain(X, y, sample_domain)
+        X_source, X_target, y_source, _ = source_target_split(
+            X, y, sample_domain=sample_domain
+        )
+        self.X_source_, self.y_source_ = X_source, y_source
+
+        self.weights_ = self._weights_optimization(X_source, X_target, y_source)
+
+        return self
+
+    def adapt(self, X, y=None, sample_domain=None, **kwargs):
+        """Predict adaptation (weights, sample or labels).
+
+        Parameters
+        ----------
+        X : array-like, shape (n_samples, n_features)
+            The source data.
+        y : array-like, shape (n_samples,)
+            The source labels.
+        sample_domain : array-like, shape (n_samples,)
+            The domain labels (same as sample_domain).
+
+        Returns
+        -------
+        X_t : array-like, shape (n_samples, n_components)
+            The data (same as X).
+        y_t : array-like, shape (n_samples,)
+            The labels (same as y).
+        weights : array-like, shape (n_samples,)
+            The weights of the samples.
+        """
+        X, sample_domain = check_X_domain(
+            X,
+            sample_domain
+        )
+
+        source_idx = extract_source_indices(sample_domain)
+
+        if source_idx.sum() > 0:
+            if np.array_equal(self.X_source_, X[source_idx]):
+                source_weights = self.weights_
+            else:
+                # get the nearest neighbor in the source domain
+                K = pairwise_kernels(
```

Comment:
why use a kernel with badwidth here? if you use the nearest neighbor then you should just use distance?